File: skills/sync-utils/scripts/check_conflicts.py
#!/usr/bin/env python3
"""
Check Conflicts - Erkennt Sync-Konflikte
"""

import sys
from pathlib import Path
from datetime import datetime

# Import sync functions
sys.path.append('/home/openclaw/.openclaw/workspace/scripts')
from sync_clawhub_git import get_file_hash

CLAWHUB_DIR = Path("/home/openclaw/.openclaw/workspace/skills")
GIT_DIR = Path("/home/openclaw/.openclaw/workspace/git/skills")

def check_conflicts():
    """Prüft auf Konflikte zwischen ClawHub und Git"""
    conflicts = []
    
    # Alle Skills die in beiden Orten existieren
    common_skills = []
    if CLAWHUB_DIR.exists() and GIT_DIR.exists():
        clawhub_skills = {d.name for d in CLAWHUB_DIR.iterdir() if d.is_dir()}
        git_skills = {d.name for d in GIT_DIR.iterdir() if d.is_dir()}
        common_skills = clawhub_skills.intersection(git_skills)
    
    print(f"Prüfe {len(common_skills)} Skills auf Konflikte...\n")
    
    for skill in sorted(common_skills):
        clawhub_path = CLAWHUB_DIR / skill
        git_path = GIT_DIR / skill
        
        # Alle Dateien vergleichen
        skill_conflicts = []
        
        # ClawHub Dateien
        clawhub_files = {}
        for f in clawhub_path.rglob('*'):
            if f.is_file() and '.git' not in str(f):
                rel_path = f.relative_to(clawhub_path)
                clawhub_files[str(rel_path)] = f
        
        # Git Dateien
        git_files = {}
        for f in git_path.rglob('*'):
            if f.is_file() and '.git' not in str(f):
                rel_path = f.relative_to(git_path)
                git_files[str(rel_path)] = f
        
        # Vergleiche gemeinsame Dateien
        for rel_path in set(clawhub_files.keys()).intersection(git_files.keys()):
            clawhub_file = clawhub_files[rel_path]
            git_file = git_files[rel_path]
            
            if get_file_hash(clawhub_file) != get_file_hash(git_file):
                clawhub_mtime = datetime.fromtimestamp(clawhub_file.stat().st_mtime)
                git_mtime = datetime.fromtimestamp(git_file.stat().st_mtime)
                
                skill_conflicts.append({
                    "file": rel_path,
                    "clawhub_modified": clawhub_mtime.strftime('%Y-%m-%d %H:%M:%S'),
                    "git_modified": git_mtime.strftime('%Y-%m-%d %H:%M:%S'),
                    "newer": "clawhub" if clawhub_mtime > git_mtime else "git"
                })
        
        if skill_conflicts:
            conflicts.append({
                "skill": skill,
                "conflicts": skill_conflicts
            })
    
    # Ausgabe
    if conflicts:
        print("⚠️  KONFLIKTE GEFUNDEN:")
        print("=" * 80)
        
        for conflict in conflicts:
            print(f"\n📦 Skill: {conflict['skill']}")
            print("-" * 40)
            
            for file_conflict in conflict['conflicts']:
                print(f"  📄 {file_conflict['file']}")
                print(f"     ClawHub: {file_conflict['clawhub_modified']}")
                print(f"     Git:     {file_conflict['git_modified']}")
                print(f"     Neuer:   {file_conflict['newer'].upper()}")
                print()
        
        print("=" * 80)
        print(f"Gesamt: {len(conflicts)} Skills mit Konflikten")
        print("\nNutze 'sync_utils/scripts/resolve_conflict.py' zum Auflösen.")
    else:
        print("✅ Keine Konflikte gefunden!")
        print("Alle gemeinsamen Skills sind synchron.")

def main():
    """Hauptfunktion"""
    check_conflicts()

if __name__ == "__main__":
    main()

File: javascript/check_conflicts.js
#!/usr/bin/env node
// check_conflicts.py — portiert nach javascript
// Quelle: python, OpenClaw@gateway1:skills/sync-utils/scripts/check_conflicts.py
// auch in: OpenClaw@gateway2:skills/sync-utils/scripts/check_conflicts.py
// Erzeugt: 2026-08-09 durch ABSTRACTIONS_MANAGER.py

/**
 * Check Conflicts - Erkennt Sync-Konflikte
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');

// Import sync functions
const { getFileHash } = require('/home/openclaw/.openclaw/workspace/scripts/sync_clawhub_git.js');

const CLAWHUB_DIR = "/home/openclaw/.openclaw/workspace/skills";
const GIT_DIR = "/home/openclaw/.openclaw/workspace/git/skills";

function getFilesRecursively(dir, baseDir) {
    let files = {};
    const items = fs.readdirSync(dir);
    
    for (const item of items) {
        const fullPath = path.join(dir, item);
        const relativePath = path.relative(baseDir, fullPath);
        
        // Skip .git directories
        if (item === '.git') continue;
        
        const stat = fs.statSync(fullPath);
        if (stat.isDirectory()) {
            const subFiles = getFilesRecursively(fullPath, baseDir);
            Object.assign(files, subFiles);
        } else {
            files[relativePath] = fullPath;
        }
    }
    
    return files;
}

function checkConflicts() {
    /** Prüft auf Konflikte zwischen ClawHub und Git */
    let conflicts = [];
    
    // Alle Skills die in beiden Orten existieren
    let commonSkills = [];
    if (fs.existsSync(CLAWHUB_DIR) && fs.existsSync(GIT_DIR)) {
        const clawhubSkills = fs.readdirSync(CLAWHUB_DIR)
            .filter(item => fs.statSync(path.join(CLAWHUB_DIR, item)).isDirectory());
        const gitSkills = fs.readdirSync(GIT_DIR)
            .filter(item => fs.statSync(path.join(GIT_DIR, item)).isDirectory());
        
        commonSkills = clawhubSkills.filter(skill => gitSkills.includes(skill));
    }
    
    console.log(`Prüfe ${commonSkills.length} Skills auf Konflikte...\n`);
    
    for (const skill of commonSkills.sort()) {
        const clawhubPath = path.join(CLAWHUB_DIR, skill);
        const gitPath = path.join(GIT_DIR, skill);
        
        // Alle Dateien vergleichen
        let skillConflicts = [];
        
        // ClawHub Dateien
        const clawhubFiles = getFilesRecursively(clawhubPath, clawhubPath);
        
        // Git Dateien
        const gitFiles = getFilesRecursively(gitPath, gitPath);
        
        // Vergleiche gemeinsame Dateien
        const commonFiles = Object.keys(clawhubFiles).filter(file => Object.keys(gitFiles).includes(file));
        
        for (const relPath of commonFiles) {
            const clawhubFile = clawhubFiles[relPath];
            const gitFile = gitFiles[relPath];
            
            if (getFileHash(clawhubFile) !== getFileHash(gitFile)) {
                const clawhubStat = fs.statSync(clawhubFile);
                const gitStat = fs.statSync(gitFile);
                
                const clawhubMtime = new Date(clawhubStat.mtime);
                const gitMtime = new Date(gitStat.mtime);
                
                skillConflicts.push({
                    "file": relPath,
                    "clawhub_modified": clawhubMtime.toISOString().slice(0, 19).replace('T', ' '),
                    "git_modified": gitMtime.toISOString().slice(0, 19).replace('T', ' '),
                    "newer": clawhubMtime > gitMtime ? "clawhub" : "git"
                });
            }
        }
        
        if (skillConflicts.length > 0) {
            conflicts.push({
                "skill": skill,
                "conflicts": skillConflicts
            });
        }
    }
    
    // Ausgabe
    if (conflicts.length > 0) {
        console.log("⚠️  KONFLIKTE GEFUNDEN:");
        console.log("=".repeat(80));
        
        for (const conflict of conflicts) {
            console.log(`\n📦 Skill: ${conflict['skill']}`);
            console.log("-".repeat(40));
            
            for (const fileConflict of conflict['conflicts']) {
                console.log(`  📄 ${fileConflict['file']}`);
                console.log(`     ClawHub: ${fileConflict['clawhub_modified']}`);
                console.log(`     Git:     ${fileConflict['git_modified']}`);
                console.log(`     Neuer:   ${fileConflict['newer'].toUpperCase()}`);
                console.log();
            }
        }
        
        console.log("=".repeat(80));
        console.log(`Gesamt: ${conflicts.length} Skills mit Konflikten`);
        console.log("\nNutze 'sync_utils/scripts/resolve_conflict.py' zum Auflösen.");
    } else {
        console.log("✅ Keine Konflikte gefunden!");
        console.log("Alle gemeinsamen Skills sind synchron.");
    }
}

function main() {
    /** Hauptfunktion */
    checkConflicts();
}

main();
